### DISCUSSION TITLE
Margin collapsing

### DISCUSSION THREAD

**Discussion by michal-perlakowski** [diegomura/react-pdf#1600] [Q&A]
Is [margin collapsing](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Box_Model/Mastering_margin_collapsing) supported? If not, is there any way to achieve a similar effect other than manually overriding margins based on siblings?

**Comment by jeetiss** [comment]
Hi @michal-perlakowski 
no, margin collapsing isn't supported, because `@react-pdf` uses `yoga-layout` (this is flexbox engine), so all elements are located with `display: flex` and don't collapse.

> is there any way to achieve a similar effect other than manually overriding margins based on siblings?

I'm not sure that I clearly understand you. Why do you need margin collapsing? Which problem you wanna solve?

**Reply by michal-perlakowski** [reply]
@jeetiss I have requirements from a UX designer that certain elements (headers, paragraphs) need to have certain margin before and/or after. But these margins can be shared between elements (i.e. collapsed).

So for example, there's:
- header1 with 20pt margin before and after
- header2 with 15pt margin before and after
- paragraph with 10pt margin before and after

Here's an example document with the correct margins:

```html
<!-- 20pt -->
<header1 />
<!-- 20pt -->
<paragraph />
<!-- 10pt -->
<paragraph />
<!-- 20pt -->
<header1 />
<!-- 20pt -->
<header2 />
<!-- 15pt -->
<paragraph />
<!-- 15pt -->
<header2 />
<!-- 15pt -->
<paragraph />
<!-- 10pt -->
```

So without margin collapsing, I need to manually specify margins based on what is the previous and the next element.

**Reply by jeetiss** [reply]
Thanks for clarification.

Yes, margin collapsing is a feature of a block formatting context, but `@react-pdf` supports only flexbox. as I say before and it will not trivial to add this.

You can use component [like this](https://react-pdf.org/repl?code=140630f60760ce02e00406502988604b49c0bc700501bc00b24043004c9009c01a39d1801b256910b41b32a42017c04a6c00f8f30387000f0035344803ba0d162e11521529c00641a2401524003de2c009e4cb3e7c8a95c00b625280733410758000e00b8e0026000cd4ac94ed1d9c0084c06060c06cbde898e001f87d7ce0bc01197d0278790455c8a879c401e8f50d047903f0e29135b5c54a8c614c91cd2dacc5829c5ddcbd89f3d49253d2015803db6decbbc323a2bd59d939b912e18752e133adb373ab0a4a0c61cb032be638b979158ba4e414f981412160e00114015cd1f4226a717004b185702e0011300805e366e01d02e200028901c3513199f088a414000740023301918c3940989c4c85406020703e9a8b0007272550d294b8029c604869c051e6147a2dcf638650486e420e4b9bc872f3f9c506a3299f57dab29a488e5a2853cbe40b044a914abc5fb497598a44f4240754a7c449f524b26a8a8549a650e974342305ad4cb651bcf4a36ea596ce45cb518aee46bf98280e8b085ab289af528034403d7048d9b30d56b4badd0c9373265de857ab43c1e1687c390f19478986a851561f0c971581a0f0441217c00371dc006290181a2b84e58150088a00198906cec63179290079282c8481c4a58d59941004f0891371403c4522a0f3be8872c048e8101a3c036229408a6019dcec845001b9a400cc450026801f474006949000d8c00007000ea603de18800aa6413e363be00068009c2400082485a29120e94b00fc2b6f73408d33450360882fa0831048176201704792003988588e25e1b44a17264190ce038ae2787013ea32042c5b110038d3144311c0bfaf1d63f1ec4001260250680005e9d9ce5e0f18b8f08bae62aa3181274ce271e906c70069584b6770004aa42a0d0a026db76dc1a8b8388af3bc9f1c045208665000), otherwise you need to count sibling and manually overriding margins based on them! 🥲
